### PR TITLE
Dockerize app

### PR DESCRIPTION
**Pull request by kareemmahlees**
## What type of Pull Request is this?
**Delete all options except for the one that applies**
- Feature

## What is the current behavior?

Issue Number: #50 

## What is the new behavior?

- An image is built for the app
- The image was provided to the docker-compose
- the app is dependent on the db

## Other information

I would like to hear your feedback if everything is okay before moving on to add this section to the CONTRIBUTING file

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,14 @@
+FROM node:latest AS development
```
Comment (omar2205) [inline Dockerfile:1]: Don't use `latest`, pin an actual version.

File: docker-compose.yml [hunk 2]
```diff
@@ -11,4 +11,19 @@ services:
     ports:
       - ${DB_PORT}:5432
     volumes:
-      - ./docker-data/db:/var/lib/postgresql/data
+      - pgdata:/var/lib/postgresql/data
```
Comment (omar2205) [inline docker-compose.yml:14]: Why did you change this?

**Review by omar2205 — CHANGES_REQUESTED** [review]

File: Dockerfile (hunk 1 above)
Comment (kareemmahlees) [reply Dockerfile]: resolved

File: docker-compose.yml (hunk 2 above)
Comment (kareemmahlees) [reply docker-compose.yml:14]: > Why did you change this?


Generally speaking there is not much difference but I decided to use a named volume to allow it to be shared with other, future, containers.
Relative volume paths are generally used when you want to sync a folder in your directory with the docker container and may cause some issues on deploying to a server, a bitter description is in this [stackoverflow question](https://stackoverflow.com/questions/46907558/docker-compose-relative-paths-vs-docker-volume#:~:text=A%20reason%20you,with%20relative%20paths.)

Comment (omar2205) [reply docker-compose.yml:14]: You should have added this to the "What is the new behaviour?" section and your reasoning.

@w3cj will have to decide if he wants a named volume.

Comment (kareemmahlees) [reply docker-compose.yml:14]: My bad!

**Comment by w3cj** [comment]
Let's focus on a production build version of the svelte app and put it in a separate docker-compose file `docker-compose.prod.yml`. This file should only have the app service and not the db service.

I've added a node adapter for the svelte build. I'll add some comments on your PR on how you can target the production build in the `Dockerfile`

I'll not that the setup for using docker in development will be a bit different, but lets do that in a separate PR after we have the production one setup.

File: Dockerfile [hunk 3]
```diff
@@ -0,0 +1,14 @@
+FROM node:18-alpine AS development
+
+WORKDIR /app
+
+COPY package*.json /app/
+
+RUN npm ci
+
+COPY . .
+
```
Comment (w3cj) [inline Dockerfile:10]: Add a command: `NODE_BUILD=true npm run build`
You can setup the environment variable `NODE_BUILD=true` on a separate line if necessary.

File: Dockerfile [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+EXPOSE 5173
```
Comment (w3cj) [inline Dockerfile:11]: The node build uses port 3000 by default

File: Dockerfile [hunk 5]
```diff
@@ -0,0 +1,14 @@
...
+
+CMD npm run docker:dev
```
Comment (w3cj) [inline Dockerfile:13]: The node build will use the command: `node build`

File: docker-compose.yml [hunk 6]
```diff
@@ -11,4 +11,19 @@ services:
...
+
+  app:
```
Comment (w3cj) [inline docker-compose.yml:16]: Like I mentioned in the PR comment, lets move this service to a separate file `docker-compose.prod.yml`

File: docker-compose.yml [hunk 7]
```diff
@@ -11,4 +11,19 @@ services:
...
+    build: .
+    depends_on:
+      - db
+    restart: always
+    ports:
+      - 5173:5173
+
+    environment:
+      AUTH_SECRET: ${AUTH_SECRET}
```
Comment (w3cj) [inline docker-compose.yml:25]: All of the environment variables will need to be exposed. See `.env.example` for a list of them.

File: package.json [hunk 8]
```diff
@@ -5,6 +5,7 @@
 	"scripts": {
 		"prepare": "husky install",
 		"dev": "concurrently \"vite dev\" typesafe-i18n",
+		"docker:dev": "concurrently \"vite dev --host 0.0.0.0\" typesafe-i18n",
```
Comment (w3cj) [inline package.json:8]: No need for an extra script here. The `node build` command will work automatically after the `build` directory is created during the build process.

**Review by w3cj — CHANGES_REQUESTED** [review]

File: .dockerignore [hunk 9]
```diff
@@ -0,0 +1,17 @@
...
+.npmrc
+.prettierignore
+.prettierrc.cjs
+.typesafe-i18n.json
+CODE_OF_CONDUCT.md
+CONTRIBUTING.md
+DEVLOG.md
+LICENSE
+NOTES.md
+README.md
```
Comment (w3cj) [inline .dockerignore:17]: I think you can safely ignore `*.md` files instead of listing them manually.

File: .dockerignore [hunk 10]
```diff
@@ -0,0 +1,17 @@
+.github
+.vscode
+.node_modules
+.husky
+.eslintignore
+.eslintrc.cjs
```
Comment (w3cj) [inline .dockerignore:6]: I'm not sure how `.dockerignore` is used exactly, but for a development container (which we'll setup later), these configuration files will be needed. They will likely be mapped to a volume that mirrors this directory.

File: Dockerfile [hunk 11]
```diff
@@ -0,0 +1,14 @@
+FROM node:18-alpine AS development
```
Comment (omar2205) [inline Dockerfile:1]: Change this to `as builder`

File: Dockerfile (hunk 5 above)
Comment (omar2205) [reply Dockerfile:13]: If we are using this for production, we can add a second stage and copy the output folder.

After the node build command add this:

```Dockerfile
FROM node:18-alpine

COPY --from=builder ./build .

ENV NODE_ENV=production

EXPOSE 3000

CMD node build

```

Comment (w3cj) [reply Dockerfile:13]: Good catch. We still need the `npm run build` command, but we only need to copy over the `build` folder, not the whole app source code.

File: .dockerignore (hunk 10 above)
Comment (kareemmahlees) [reply .dockerignore:6]: Do you mean you want to create a dev container where the users can open their vscode inside the container and work from there?

File: docker-compose.yml (hunk 7 above)
Comment (kareemmahlees) [reply docker-compose.yml:25]: > All of the environment variables will need to be exposed. See `.env.example` for a list of them.

I am not really sure I understand what you mean by exposed ?!

File: .dockerignore (hunk 10 above)
Comment (omar2205) [reply .dockerignore:6]: Two containers (docker-compose) files, dev and prod versions.

Comment (kareemmahlees) [reply .dockerignore:6]: > Two containers (docker-compose) files, dev and prod versions.

I did that in the latest commit, but I kept the configuration files ignored in `.dockerignore` and every thing seems to run ok!

**Comment by w3cj** [comment]
There are several changes I need you to make, so lets just focus on the production `Dockerfile` and `docker-compose.prod.yml` for now.

Remove `docker-compose.dev.yml` and `.dockerignore`, we will add these in a future PR where we can discuss the development container specifically.

`Dockerfile`:
1. Should run the command `npm run build`
2. Should copy files from the generated folder called `build` instead of the whole app
3. Should expose port `3000`
4. Should run the command `node build`

`docker-compose.prod.yml`:
1. Remove `depends_on: - db` - the production container will connect to a remote DB
2. Every environment variable listed in the [.env.example](https://github.com/CodingGarden/listd/blob/main/.env.example) file should be listed under `environment`

Remove current changes to the `docker-compose.yml` file

**Review by w3cj — CHANGES_REQUESTED** [review]
See comment above.

**Comment by kareemmahlees** [comment]
@w3cj sorry for being late, was busy lately.
i did all the changes mentioned here:
> There are several changes I need you to make, so lets just focus on the production Dockerfile and docker-compose.prod.yml for now.

but regarding this part:
>Every environment variable listed in the [.env.example](https://github.com/CodingGarden/listd/blob/main/.env.example) file should be listed under environment

couldn't we just do this:
```yaml
 env_file:
    - .env
```

**Comment by w3cj** [comment]
Thanks for this. Looks good, but the original `docker-compose.yml` file got deleted. Be sure to leave the original as well as the new one you created.

**Comment by w3cj** [comment]
> 

 I think we want to avoid doing this, because the production container likely won't have a `.env` file. The variables will already be loaded in the system or specified when launching the container.

**Comment by kareemmahlees** [comment]
@w3cj Done, could you check?

File: Dockerfile [hunk 12]
```diff
@@ -0,0 +1,27 @@
...
+
+ENV NODE_BUILD=true
+
+RUN npm run build
+
+EXPOSE 3000
+
+CMD node build
+
+FROM node:18-alpine as production
```
Comment (w3cj) [inline Dockerfile:19]: Is there a reason to split the build and run into separate containers?

Comment (w3cj) [reply Dockerfile:19]: Ok seems fine then, but does there need to be an EXPOSE and CMD for the first stage? Isn't it needed for only the second container?

Comment (omar2205) [reply Dockerfile:19]: Yeah, @kareemmahlees just remove the `EXPOSE` and `CMD node build` from the build stage.
